src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -163,5 +176,107 @@ private IEnumerable<CodeRefactoringProvider> GetProviders(Document document)
 
             return null;
         }
+
+        private ImmutableArray<CodeRefactoringProvider> GetProjectRefactorings(Project project)
+        {
+            // TODO (https://github.com/dotnet/roslyn/issues/4932): Don't restrict refactorings in Interactive
+            return project.Solution.Workspace.Kind == WorkspaceKind.Interactive
+                ? ImmutableArray<CodeRefactoringProvider>.Empty
+                : _projectRefactoringsMap.GetValue(project.AnalyzerReferences, pId => new StrongBox<ImmutableArray<CodeRefactoringProvider>>(ComputeProjectRefactorings(project))).Value;
+        }
+
+        private ImmutableArray<CodeRefactoringProvider> ComputeProjectRefactorings(Project project)
+        {
+            var extensionManager = project.Solution.Workspace.Services.GetService<IExtensionManager>();
+            ImmutableArray<CodeRefactoringProvider>.Builder builder = null;
+            foreach (var reference in project.AnalyzerReferences)
+            {
+                var projectCodeRefactoringProvider = _analyzerReferenceToRefactoringsMap.GetValue(reference, _createProjectCodeRefactoringsProvider);
+                foreach (var refactoring in projectCodeRefactoringProvider.GetRefactorings(project.Language))
+                {
+                    builder ??= ImmutableArray.CreateBuilder<CodeRefactoringProvider>();
+                    builder.Add(refactoring);
+                }
+            }
+
+            if (builder is null)
+            {
+                return ImmutableArray<CodeRefactoringProvider>.Empty;
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private class ProjectCodeRefactoringProvider
```

Comment:
nit: consider extracting to its own file.
nit: i'm on the fence if 'Project' or 'AnalyzerReference' is preferred in the name here.  I kinda like AnalyzerReferenceCodeRefactoringsProvider.

Reply:
Copied the pattern from `CodeFixService`. One option is to make a follow-up to change both to something else.

Reply:
I think we extract most of the common functionality between this class and [ProjectCodeFixProvider](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/CodeFixes/CodeFixService.ProjectCodeFixProvider.cs,17) into a common generic base type.

Reply:
Indeed. looks very shareable.

Reply:
Leaving this for a follow-up